src/servers/graphql-server.ts

```diff
@@ -320,77 +346,167 @@ export const startApolloServer = async ({
 
   apolloServer.applyMiddleware({ app, path: "/graphql" })
 
+  // old legacy ws
+  const onConnect = async (
+    connectionParams: Record<string, unknown>,
+    webSocket: unknown,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    connectionContext: any,
+  ) => {
+    const { request } = connectionContext
+
+    const authz = (connectionParams.authorization || connectionParams.Authorization) as
+      | string
+      | undefined
+
+    // TODO: also manage the case where there is a cookie in the request
+    // https://www.ory.sh/docs/oathkeeper/guides/proxy-websockets#configure-ory-oathkeeper-and-ory-kratos
+    const cookies = request.headers.cookie
+    if (cookies?.includes("ory_kratos_session")) {
+      const kratosCookieRes = await validateKratosCookie(cookies)
+      if (kratosCookieRes instanceof Error) return kratosCookieRes
+      const tokenPayload = {
+        sub: kratosCookieRes.kratosUserId,
+      }
+      return sessionContext({
+        tokenPayload,
+        ip: request?.socket?.remoteAddress,
+        body: null,
+      })
+    }
+
+    // make request to oathkeeper
+    const originalToken = authz?.slice(7) as LegacyJwtToken | SessionToken | undefined
+
+    const newToken = await sendOathkeeperRequest(originalToken)
+    // TODO: see how returning an error affect the websocket connection
+    if (newToken instanceof Error) return newToken
+
+    const keyJwks = await jwksRsa(getJwksArgs()).getSigningKey()
+
+    const tokenPayload = jsonwebtoken.verify(newToken, keyJwks.getPublicKey(), {
+      algorithms: jwtAlgorithms,
+    })
+
+    if (typeof tokenPayload === "string") {
+      throw new Error("tokenPayload should be an object")
+    }
+
+    return sessionContext({
+      tokenPayload,
+      ip: request?.socket?.remoteAddress,
+
+      // TODO: Resolve what's needed here
+      body: null,
+    })
+  }
+
+  // new ws server
+  /* eslint @typescript-eslint/ban-ts-comment: "off" */
+  // @ts-ignore-next-line no-implicit-any error
+  const context = async (ctx) => {
+    const connectionParams = ctx.connectionParams as Record<string, string>
+
+    // TODO: check if nginx pass the ip to the header
+    // TODO: ip not been used currently for subscription.
+    // implement some rate limiting.
+    const ipString = isDev
+      ? connectionParams?.ip
+      : connectionParams?.["x-real-ip"] || connectionParams?.["x-forwarded-for"]
+
+    const ip = parseIps(ipString)
+
+    const authz = (connectionParams.authorization || connectionParams.Authorization) as
+      | string
+      | undefined
+
+    // TODO: also manage the case where there is a cookie in the request
```

Comment:
Is this a copy of the above code at line 362?

Reply:
yes, I want to sync with @ntheile on this